apps/taxes/src/modules/taxes/tax-provider-utils.ts

```diff
@@ -1,25 +1,34 @@
 import { z } from "zod";
+import { TaxUnexpectedError } from "./tax-error";
 
 /*
  * The providers sdk types claim to sometimes return undefined.
  * If it ever happens, we have nothing to fall back to, so we throw an error.
  * Should only be used for values that are required for further calculation.
  */
-function resolveOptionalOrThrow<T>(value: T | undefined | null, error?: Error): T {
+function resolveOptionalOrThrow<T>(value: T | undefined | null, errorMessage?: string): T {
   if (value === undefined || value === null) {
-    throw error
-      ? error
-      : new Error("Could not resolve data. Value needed for further calculation is undefined.");
+    throw new TaxUnexpectedError(
+      errorMessage
+        ? errorMessage
+        : "Could not resolve data. Value needed for further calculation is undefined.",
+    );
   }
 
   return value;
 }
 
 function resolveStringOrThrow(value: string | undefined | null): string {
-  return z
+  const parseResult = z
     .string({ required_error: "This field must be defined." })
     .min(1, { message: "This field can not be empty." })
-    .parse(value);
+    .safeParse(value);
+
+  if (!parseResult.success) {
+    throw new TaxUnexpectedError(parseResult.error.message);
```

Comment:
How is it unexpected? Every time you pass wrong data to this function, validator will fail. This is very much expected. 

If this function is used in specific case, where this should not happen, then it may be unexpected, because of the  context

Reply:
I realized the naming problem was in trying to generalize different types of errors. I replaced `TaxUnexpectedError` with two errors: `TaxIncompletePayloadError` and `TaxIncompleteResponseError`.